cpp/benchmarks/string/contains.cpp

```diff
@@ -28,39 +30,77 @@ class StringContains : public cudf::benchmark {
+  // For hit-rate ~10%, matches is 10 -- use all rows from data (only 10% will match).
+  auto matches = static_cast<int32_t>(data_view.size() * rawdata.size() / hit_rate);
+  data_profile table_profile;
+  table_profile.set_distribution_params(
+    cudf::type_id::INT32, distribution_id::UNIFORM, 0, matches - 1);
+  table_profile.set_null_frequency(0.0);  // no nulls for gather-map
+
+  auto table      = create_random_table({cudf::type_id::INT32}, row_count{n_rows}, table_profile);
+  auto gather_map = cudf::column_view(table->view().column(0));
+  table           = cudf::gather(cudf::table_view({data_view}), gather_map);
+  auto input      = cudf::strings_column_view(table->view().column(0));
+
+  auto pattern = patterns[pattern_index];
 
   for (auto _ : state) {
     cuda_event_timer raii(state, true, rmm::cuda_stream_default);
-    // contains_re(), matches_re(), and count_re() all have similar functions
-    // with count_re() being the most regex intensive
     switch (ct) {
       case contains_type::contains:  // contains_re and matches_re use the same main logic
-        cudf::strings::contains_re(input, "\\d+");
+        cudf::strings::contains_re(input, pattern);
         break;
-      case contains_type::count:  // counts occurrences of pattern
-        cudf::strings::count_re(input, "\\d+");
+      case contains_type::count:  // counts occurrences of matches
+        cudf::strings::count_re(input, pattern);
         break;
-      case contains_type::findall:  // returns occurrences of matches
-        cudf::strings::findall(input, "\\d+");
+      case contains_type::findall:  // returns occurrences of all matches
+        cudf::strings::findall(input, pattern);
         break;
     }
   }
 
   state.SetBytesProcessed(state.iterations() * input.chars_size());
 }
 
-#define STRINGS_BENCHMARK_DEFINE(name, b)                          \
-  BENCHMARK_DEFINE_F(StringContains, name)                         \
-  (::benchmark::State & st) { BM_contains(st, contains_type::b); } \
-  BENCHMARK_REGISTER_F(StringContains, name)                       \
-    ->RangeMultiplier(8)                                           \
-    ->Ranges({{1 << 12, 1 << 24}})                                 \
-    ->UseManualTime()                                              \
+#define STRINGS_BENCHMARK_DEFINE(name, b)                                                   \
+  BENCHMARK_DEFINE_F(StringContains, name)                                                  \
+  (::benchmark::State & st) { BM_contains(st, contains_type::b); }                          \
+  BENCHMARK_REGISTER_F(StringContains, name)                                                \
+    ->ArgsProduct({{4096, 32768, 262144, 2097152, 16777216}, /* row count */                \
+                   {0, 1},                                   /* patterns index */           \
+                   {10, 18, 46, 64, 98}})                    /* 11%, 25%, 50%, 70%, 100% */ \
```

Comment:
I was not sure the best way to communicate these numbers. There were empirically found by trying all values between 10-100 and seeing which ones correspond closest to the desired hit-rates in the results.

Reply:
Can we adjust the way the random data is generated? I have read it through 3-4 times and I'm still not sure how it is supposed to work. It would make sense to me for the algorithm to do the following:
1. Given a `hit_rate` (float between 0.0 and 1.0).
2. Draw `target_size` random values (floats between 0.0 and 1.0).
3. If `hit_rate < random_values[i]`, draw a random string from a set of "matches," else draw a random string from a set of "non-matches."
4. Assemble the target column from the randomly selected matches and non-matches. The proportion of matches should be approximately `hit_rate`.

One possible implementation (in NumPy for ease of use, but not too tough to port to Thrust):
```python
import numpy as np

raw_data = np.asarray([
    "123 abc 4567890 DEFGHI 0987 5W43",  # matches both patterns;
    "012345 6789 01234 56789 0123 456",  # the rest do not match
    "abc 4567890 DEFGHI 0987 Wxyz 123",
    "abcdefghijklmnopqrstuvwxyz 01234",
    "",
    "AbcéDEFGHIJKLMNOPQRSTUVWXYZ 01",
    "9876543210,abcdefghijklmnopqrstU",
    "9876543210,abcdefghijklmnopqrstU",
    "123 édf 4567890 DéFG 0987 X5",
    "1",
])

num_matches = 1
num_non_matches = len(raw_data) - num_matches

target_size = 10
hit_rate = 0.1
match_indices = np.random.randint(num_matches, size=target_size)
non_match_indices = np.random.randint(num_non_matches, size=target_size)
hit = np.random.rand(target_size)
gather_map = np.where(hit < hit_rate, match_indices, num_matches + non_match_indices)
print(raw_data[gather_map])
```

Reply:
The `hit_rate` is meant to be a percentage (1-100%). Although this is certainly possible with float values (0.01 to 1.0) I don't see a need for this precision -- I don't need 43.2% or 0.1%. My goal was to make these 10%, 25%, 50%, 70%, 100% as per the comment but the actual randomized (uniform) results were not uniform enough and so I adjusted the numbers to match the actual hit-rate I was targeting.

I will look at building a non-randomized gather-map so the numbers can match the actual target results.

Reply:
I used some of @bdice approach above to build a full gather-map of all non-matched indices. Then I used the `cudf::sequence` API to build a uniform scatter-map that scatters the matched index (idx=0) throughout the gather-map. This will give a more precise spread of matches based on the hit-rate percentage value. Hopefully this will be easier to follow.